pulp_ansible/app/serializers.py

```diff
@@ -361,151 +373,304 @@ class Meta:
+    A serializer with the logic necessary to upload a CollectionVersion.
+
+    Used in ``.viewsets.CollectionVersionViewSet`` and ``.galaxy.v3.views.CollectionUploadViewSet``
+    to perform the creation and validation of a CollectionVersion and add to repository if
+    necessary. This serializer is meant to be compliant with
+    ``pulpcore.plugin.viewsets.SingleArtifactContentUploadViewSet`` and thus follows these steps on
+    creation:
+
+    1. ``SingleArtifactContentUploadViewSet.create()`` converts uploaded-file (if present) to an
+        artifact and calls ``validate()`` with request in context
+    2. ``pulpcore.plugin.tasks.general_create`` is dispatched with this serializer and its initial
+        validated data and a deferred context determine by the viewset (default an empty context).
+    3. ``general_create`` calls ``validate()`` again with deferred context now.
+        ``deferred_validate()`` is now called and the upload object (if present) is converted to an
+        artifact.
+    4. ``general_create`` calls ``save()`` which will call ``create()``. ``create`` uses the
+        validated data to create and save the CollectionVersion. If repository is specified the
+        CollectionVersion is then added to the repository.
     """
 
-    name = serializers.CharField(help_text=_("The name of the collection."), max_length=64)
+    sha256 = serializers.CharField(
+        help_text=_("An optional sha256 checksum of the uploaded file."),
+        required=False,
+        write_only=True,
+    )
 
-    namespace = serializers.CharField(
-        help_text=_("The namespace of the collection."), max_length=64
+    expected_name = serializers.CharField(
+        help_text=_("The name of the collection."),
+        max_length=64,
+        required=False,
+        write_only=True,
     )
 
-    version = serializers.CharField(help_text=_("The version of the collection."), max_length=128)
+    expected_namespace = serializers.CharField(
+        help_text=_("The namespace of the collection."),
+        max_length=64,
+        required=False,
+        write_only=True,
+    )
+
+    expected_version = serializers.CharField(
+        help_text=_("The version of the collection."),
+        max_length=128,
+        required=False,
+        write_only=True,
+    )
 
     def validate(self, data):
-        """Validate that we have a file or can create one."""
-        if "artifact" in data:
-            raise serializers.ValidationError(_("Only 'file' may be specified."))
+        """Check and set the namespace, name & version."""
+        # This validation is for creating CollectionVersions
+        if self.instance:
+            return data
+        # Super will call deferred_validate on second call
+        data = super().validate(data)
 
-        if "request" not in self.context:
-            data = self.deferred_validate(data)
+        # First time calling validate
+        if "request" in self.context:
```

Comment:
You are basically inverting the "deferred_validate" logic here. Can we just start with the raw data again in the task, so we can be sure that we validate everything in one go?

Reply:
Yes, I can rewrite it. Just need to take into account not having the artifact filename on the second validate.